runtime/src/message_processor.rs

```diff
@@ -302,22 +303,28 @@ impl<'a> ThisInvokeContext<'a> {
     }
 }
 impl<'a> InvokeContext for ThisInvokeContext<'a> {
-    fn push(&mut self, key: &Pubkey) -> Result<(), InstructionError> {
-        if self.program_ids.len() > self.bpf_compute_budget.max_invoke_depth {
+    fn push(
+        &mut self,
+        keyed_accounts: &[KeyedAccount],
+        key: &Pubkey,
+    ) -> Result<&[KeyedAccount], InstructionError> {
+        if self.invoke_stack.len() > self.bpf_compute_budget.max_invoke_depth {
             return Err(InstructionError::CallDepth);
         }
-        if self.program_ids.contains(key) && self.program_ids.last() != Some(key) {
+        let frame_index = self.invoke_stack.iter().position(|frame| frame.0 == *key);
+        if frame_index != None && frame_index != Some(self.invoke_stack.len() - 1) {
             // Reentrancy not allowed unless caller is calling itself
             return Err(InstructionError::ReentrancyNotAllowed);
         }
-        self.program_ids.push(*key);
-        Ok(())
+        let keyed_accounts = unsafe { std::mem::transmute(keyed_accounts) };
```

Comment:
Gotta get rid of this, maybe we need to rethink keyed_account, those direct references have always been a pain.  Maybe `Rc` would help here?

Reply:
I think we can just put the `KeyedAccount`s in the `InvokeContext` directly, without a reference. As it is now the only place where it resides.